modules/collect/src/main/java/com/opengamma/strata/collect/timeseries/DenseLocalDateDoubleTimeSeries.java

```diff
@@ -215,41 +222,53 @@ static LocalDateDoubleTimeSeries of(
 
     double[] points = new double[dateCalculation.calculatePosition(startDate, endDate) + 1];
     Arrays.fill(points, Double.NaN);
-    values.forEach(pt -> points[dateCalculation.calculatePosition(startDate, pt.getDate())] = pt.getValue());
-    return new DenseLocalDateDoubleTimeSeries(startDate, points, dateCalculation, true);
+    int size = 0;
+    for (LocalDateDoublePoint pt : Guavate.in(values)) {
+      points[dateCalculation.calculatePosition(startDate, pt.getDate())] = pt.getValue();
+      size++;
+    }
+    return new DenseLocalDateDoubleTimeSeries(startDate, points, dateCalculation, true, size);
   }
 
   // Private constructor, the trusted flag indicates whether the
   // points array should be cloned. If trusted, it will not be cloned.
+  // size is the size of the time series if known, -1 if unknown
   private DenseLocalDateDoubleTimeSeries(
       LocalDate startDate,
       double[] points,
       DenseTimeSeriesCalculation dateCalculation,
-      boolean trusted) {
+      boolean trusted,
+      int size) {
 
     ArgChecker.notNull(points, "points");
     this.startDate = ArgChecker.notNull(startDate, "startDate");
     this.points = trusted ? points : points.clone();
     this.dateCalculation = ArgChecker.notNull(dateCalculation, "dateCalculation");
+    this.size = size; 
   }
 
   @ImmutableConstructor
   private DenseLocalDateDoubleTimeSeries(
       LocalDate startDate,
       double[] points,
       DenseTimeSeriesCalculation dateCalculation) {
-    this(startDate, points, dateCalculation, false);
+    this(startDate, points, dateCalculation, false, -1);
   }
 
   //-------------------------------------------------------------------------
   @Override
   public boolean isEmpty() {
-    return !validIndices().findFirst().isPresent();
+    return size() == 0;
   }
 
   @Override
   public int size() {
-    return (int) validIndices().count();
+    int s = size;
```

Comment:
Add comment "threadsafe via racy single check" or similar

Reply:
@jodastephen Just thinking - is this correct if we're using `-1` as the unset value rather than `0`? I can't seem to find documentation on whether one will always see a "set" value for the field or whether it can see the "unset value" of 0 before we would set it to -1.

Reply:
True. I've never seen a case with -1 as the default. 0 has special meaning in the JVM, so best to stick with 0 as the default and adjust the size to cope (add one or size 0 -> -1)

Reply:
Yeah that was my thought, just do an offset by 1!